novawallet/Modules/Notifications/Management/NotificationsManagementInteractor.swift

```diff
@@ -0,0 +1,104 @@
+import UIKit
+import RobinHood
+import SoraKeystore
+
+final class NotificationsManagementInteractor: AnyProviderAutoCleaning {
+    weak var presenter: NotificationsManagementInteractorOutputProtocol?
+    let settingsLocalSubscriptionFactory: SettingsLocalSubscriptionFactoryProtocol
+    let settingsManager: SettingsManagerProtocol
+    let alertServiceFactory: Web3AlertsServicesFactoryProtocol
+
+    private var settingsProvider: StreamableProvider<LocalPushSettings>?
+    private var syncService: Web3AlertsSyncServiceProtocol?
+    private var pushService: PushNotificationsServiceProtocol?
+
+    init(
+        settingsLocalSubscriptionFactory: SettingsLocalSubscriptionFactoryProtocol,
+        settingsManager: SettingsManagerProtocol,
+        alertServiceFactory: Web3AlertsServicesFactoryProtocol
+    ) {
+        self.settingsLocalSubscriptionFactory = settingsLocalSubscriptionFactory
+        self.settingsManager = settingsManager
+        self.alertServiceFactory = alertServiceFactory
+    }
+
+    private func subscribeToSettings() {
+        clear(streamableProvider: &settingsProvider)
+        settingsProvider = subscribeToPushSettings()
+    }
+
+    private func provideAnnouncementsFlag() {
+        let isEnabled = settingsManager.announcements
+        DispatchQueue.main.async {
+            self.presenter?.didReceive(announcementsEnabled: isEnabled)
+        }
+    }
+
+    private func provideNotificationsStatus() {
+        if pushService == nil {
+            pushService = alertServiceFactory.createPushNotificationsService()
+        }
+
+        pushService?.statusObservable.addObserver(with: self, sendStateOnSubscription: true) { [weak self] _, status in
+            guard let status = status else {
+                return
+            }
+            DispatchQueue.main.async {
+                self?.presenter?.didReceive(notificationsEnabled: status == .on)
+            }
+        }
+    }
+}
+
+extension NotificationsManagementInteractor: NotificationsManagementInteractorInputProtocol {
+    func setup() {
+        subscribeToSettings()
+        provideAnnouncementsFlag()
+        provideNotificationsStatus()
+    }
+
+    func enableNotifications() {
+        if pushService?.statusObservable.state == .denied {
+            DispatchQueue.main.async {
+                self.presenter?.didReceive(error: .notificationsDisabledInSettings)
+            }
+        }
+    }
+
+    func save(
+        settings: LocalPushSettings,
+        notificationsEnabled: Bool,
+        announcementsEnabled: Bool
+    ) {
+        if syncService == nil {
+            syncService = alertServiceFactory.createSyncService()
+        }
+
+        syncService?.save(settings: settings) { [weak self] in
+            self?.settingsManager.notificationsEnabled = notificationsEnabled
+            self?.settingsManager.announcements = announcementsEnabled
+            DispatchQueue.main.async {
```

Comment:
Don't forget to fix queues in all places when you make the syncService always dispatch a closure in a particular queue passed in parameters